core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -185,7 +189,21 @@
     this.nameResolver.start(new NameResolver.Listener() {
       @Override
       public void onUpdate(List<ResolvedServerInfo> servers, Attributes config) {
-        loadBalancer.handleResolvedAddresses(servers, config);
+        ArrayList<ResolvedServerInfo> effectiveServers = new ArrayList<ResolvedServerInfo>();
+        for (ResolvedServerInfo server : servers) {
+          SocketAddress address = server.getAddress();
+          if (address instanceof InetSocketAddress) {
+            InetSocketAddress inetSocketAddress = (InetSocketAddress) address;
+            if (inetSocketAddress.getPort() == 0) {
+              server = new ResolvedServerInfo(
+                  new InetSocketAddress(
+                      inetSocketAddress.getAddress(), ManagedChannelImpl.this.defaultPort),
+                  server.getAttributes());
+            }
+          }
```

Comment:
This is not ideal. Based on @ejona86's suggestion, I have created #1158 and will abandon this one.